types/errors.go

```diff
@@ -0,0 +1,261 @@
+}
+
+// HandleServerError handles separate server errors and sends appropriate responses
+func HandleServerError(writer http.ResponseWriter, err error) {
+	log.Error().Err(err).Msg("handleServerError()")
+
+	var respErr error
+
+	switch err := err.(type) {
+	case *RouterMissingParamError, *RouterParsingError, *json.SyntaxError, *NoBodyError:
+		respErr = responses.SendBadRequest(writer, err.Error())
+	case *json.UnmarshalTypeError:
+		respErr = responses.SendBadRequest(writer, "bad type in json data")
+	case *ItemNotFoundError:
+		respErr = responses.SendNotFound(writer, err.Error())
+	case *AuthenticationError:
+		respErr = responses.SendForbidden(writer, err.Error())
+	default:
+		respErr = responses.SendInternalServerError(writer, "Internal Server Error")
+	}
+
+	if respErr != nil {
+		log.Error().Err(respErr).Msg(responseDataError)
+	}
+}
+
+// ErrOldReport is an error returned if a more recent already
+// exists on the storage while attempting to write a report for a cluster.
+var ErrOldReport = errors.New("More recent report already exists in storage")
+
+// ItemNotFoundError shows that item with id ItemID wasn't found in the storage
+type ItemNotFoundError struct {
+	ItemID interface{}
+}
+
+// Error returns error string
+func (e *ItemNotFoundError) Error() string {
+	return fmt.Sprintf("Item with ID %+v was not found in the storage", e.ItemID)
+}
+
+// TableNotFoundError table not found error
+type TableNotFoundError struct {
+	tableName string
+}
+
+// Error returns error string
+func (err *TableNotFoundError) Error() string {
+	return fmt.Sprintf("no such table: %v", err.tableName)
+}
+
+// TableAlreadyExistsError represents table already exists error
+type TableAlreadyExistsError struct {
+	tableName string
+}
+
+// Error returns error string
+func (err *TableAlreadyExistsError) Error() string {
+	return fmt.Sprintf("table %v already exists", err.tableName)
+}
+
+// ForeignKeyError something violates foreign key error
+// tableName and foreignKeyName can be empty for DBs not supporting it (SQLite)
+type ForeignKeyError struct {
+	TableName      string
+	ForeignKeyName string
+
+	// Details can reveal you information about specific item violating fk
+	Details string
+}
+
+// Error returns error string
+func (err *ForeignKeyError) Error() string {
+	return fmt.Sprintf(
+		`operation violates foreign key "%v" on table "%v"`, err.ForeignKeyName, err.TableName,
+	)
+}
+
+// ConvertDBError converts sql errors to those defined in this package
+func ConvertDBError(err error, itemID interface{}) error {
+	if err == nil {
+		return nil
+	}
+
+	if err == sql.ErrNoRows {
+		if itemIDArray, ok := itemID.([]interface{}); ok {
+			var strArray []string
+			for _, item := range itemIDArray {
+				strArray = append(strArray, fmt.Sprint(item))
+			}
+
+			itemID = strings.Join(strArray, "/")
+		}
+
+		return &ItemNotFoundError{ItemID: itemID}
+	}
+
+	err = convertPostgresError(err)
+	err = convertSQLiteError(err)
+
+	return err
+}
```

Comment:
This function seems to be very specific to insights-results-aggregator, as the rest of microservices doesn't have database. I'm not sure if it can be kept on insights-results-aggregator repository

Reply:
Hm, yes, probably I moved too much :D

Reply:
Fixed it there